src/compute/src/sink/persist_sink.rs

```diff
@@ -1181,3 +1190,86 @@ where
+        let mut buffer = Vec::new();
+        move |frontiers| {
+            let ac = frontiers.into_element();
+            match ac.frontier().to_owned().into_option() {
+                Some(ts) => {
+                    match refresh_schedule.round_up_timestamp(ts) {
+                        Some(rounded_up_ts) => {
+                            capability.as_mut().unwrap().downgrade(&rounded_up_ts);
+                        }
+                        None => {
+                            // We are past the last refresh. Drop the capability to signal that we
+                            // are done.
+                            capability = None;
+                        }
+                    }
+                }
+                None => {
+                    capability = None;
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
+                    data.swap(&mut buffer);
+                    let mut cached_ts: Option<Timestamp> = None;
+                    let mut cached_rounded_up_ts = None;
+                    buffer
+                        .drain_filter_swapping(|(_d, ts, _r)| {
+                            let rounded_up_ts = {
+                                // We cache the rounded up timestamp for the last seen timestamp,
+                                // because the rounding up has a non-negligible cost. Caching for
+                                // just the 1 last timestamp helps already, because in some common
+                                // cases, we'll be seeing a lot of identical timestamps, e.g.,
+                                // during a rehydration, or when we have much more than 1000 records
+                                // during a single second.
+                                let some_ts_clone = Some(ts.clone());
+                                if cached_ts != some_ts_clone {
+                                    cached_ts = some_ts_clone;
+                                    cached_rounded_up_ts = refresh_schedule.round_up_timestamp(*ts);
+                                }
+                                cached_rounded_up_ts
+                            };
+                            match rounded_up_ts {
+                                Some(rounded_up_ts) => {
+                                    *ts = rounded_up_ts;
+                                    false
+                                }
+                                None => {
+                                    // This record is after the last refresh, so drop it.
+                                    true
+                                }
+                            }
+                        })
+                        .for_each(drop); // consume the iterator to apply the filtering
+                    output_buf
+                        .activate()
+                        .session(&capability)
+                        .give_container(&mut buffer);
```

Comment:
`give_iterator` if the buffer might be not 100% filled.

Reply:
I did some refactoring and now there is a `consolidate_updates` and then a `shrink_to_fit` on the `buffer` (after the `retain_mut`). So `give_container` is ok in this case, right?

Reply:
Instead of consolidating here you might be better off using the consolidate buffer! https://github.com/MaterializeInc/materialize/blob/3ada8fb079c5e6b36e2cf1c5b0de1a296f3310a6/src/timely-util/src/buffer.rs#L41

Reply:
Ok, I switched to `ConsolidateBuffer`, could you please check? I added a method to `ConsolidateBuffer`.

255ec4bb87da544f318792b93f3cd79cb7a5bd76

Reply:
Discussed on Slack.